rpcs3/Emu/Cell/lv2/sys_rsx.cpp

```diff
@@ -548,30 +613,34 @@ error_code sys_rsx_context_attribute(u32 context_id, u32 package_id, u64 a3, u64
 		// seems gcmSysWaitLabel uses this offset, so lets set it to 0 every flip
 		vm::_ref<u32>(render->label_addr + 0x10) = 0;
 
-		//if (a3 == 0)
-		//	sys_event_port_send(rsx_cfg->rsx_event_port, 0, (1 << 3), 0);
-		//if (a3 == 1)
-		sys_event_port_send(rsx_cfg->rsx_event_port, 0, (1 << 4), 0);
+		rsx_cfg->send_event(0, SYS_RSX_EVENT_FLIP_BASE << 1, 0);
 		break;
 
 	case 0xFED: // hack: vblank command
+	{
+		// NOTE: There currently seem to only be 2 active heads on PS3
+		verify(HERE), a3 < 2;
+
 		// todo: this is wrong and should be 'second' vblank handler and freq, but since currently everything is reported as being 59.94, this should be fine
 		vm::_ref<u32>(render->device_addr + 0x30) = 1;
 		driverInfo.head[a3].vBlankCount++;
 		driverInfo.head[a3].lastSecondVTime = rsxTimeStamp();
-		sys_event_port_send(rsx_cfg->rsx_event_port, 0, (1 << 1), 0);
+
+		u64 event_flags = SYS_RSX_EVENT_VBLANK;
 
 		if (render->enable_second_vhandler)
-			sys_event_port_send(rsx_cfg->rsx_event_port, 0, (1 << 11), 0); // second vhandler
+			event_flags |= SYS_RSX_EVENT_SECOND_VBLANK_BASE << a3; // second vhandler
 
+		rsx_cfg->send_event(0, event_flags, 0);
```

Comment:
Should we really wait unconditionally for flip events? What happens if the game decides to take its time responding?
Flip code is intentionally not PS3-accurate for performance reasons, e.g to allow unlocked fps which is usually not how console works.

Reply:
The thread which pops the events is gcm interrupts thread which does it all the time (its name is bacause it takes care of those interrupts/events), theres no deadlock here.